vowpalwabbit/core/src/reductions/cb/cb_explore_adf_graph_feedback.cc

```diff
@@ -0,0 +1,549 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "vw/core/reductions/cb/cb_explore_adf_graph_feedback.h"
+
+#include "vw/common/random.h"
+#include "vw/config/options.h"
+#include "vw/core/gd_predict.h"
+#include "vw/core/global_data.h"
+#include "vw/core/label_dictionary.h"
+#include "vw/core/label_parser.h"
+#include "vw/core/model_utils.h"
+#include "vw/core/parser.h"
+#include "vw/core/reductions/cb/cb_adf.h"
+#include "vw/core/reductions/cb/cb_explore.h"
+#include "vw/core/reductions/cb/cb_explore_adf_common.h"
+#include "vw/core/setup_base.h"
+#include "vw/core/vw_math.h"
+#include "vw/explore/explore.h"
+
+#include <algorithm>
+#define ARMA_DONT_USE_BLAS
+#define ARMA_DONT_USE_LAPACK
+#include <armadillo>
+#include <cmath>
+#include <ensmallen.hpp>
+#include <functional>
+#include <vector>
+
+using namespace VW::cb_explore_adf;
+
+namespace VW
+{
+namespace cb_explore_adf
+{
+class cb_explore_adf_graph_feedback
+{
+public:
+  cb_explore_adf_graph_feedback(float gamma_scale, float gamma_exponent, VW::workspace* all)
+      : _gamma_scale(gamma_scale), _gamma_exponent(gamma_exponent), _all(all)
+  {
+  }
+  // Should be called through cb_explore_adf_base for pre/post-processing
+  void predict(VW::LEARNER::learner& base, multi_ex& examples);
+  void learn(VW::LEARNER::learner& base, multi_ex& examples);
+  void save_load(io_buf& io, bool read, bool text);
+  size_t _counter = 0;
+  float _gamma_scale;
+  float _gamma_exponent;
+
+private:
+  VW::workspace* _all;
+  template <bool is_learn>
+  void predict_or_learn_impl(VW::LEARNER::learner& base, multi_ex& examples);
+  void update_example_prediction(multi_ex& examples);
+};
+}  // namespace cb_explore_adf
+
+/**
+ * Implementing the constrained optimization from the paper: https://arxiv.org/abs/2302.08631
+ * We want to minimize objective: p * fhat + z
+ * where p is the resulting probability distribution, fhat is the scores we get from cb_adf and z is a constant
+ * The constraints are (very roughly) for each action we want:
+ *        (p - ea)**2 / G * p <= fhat_a + z
+ * where fhat_a is the cost for that action and ea is the vector of the identity matrix at index a
+ *
+ * G is a (num_actions x num_actions) graph where each column indicates the probability of that action (corresponding to
+ * the column index) revealing the reward for a different action (which row are we talking about)
+ *
+ * So for example the identity matrix for a 2 x 2 matrix
+ * 1 0
+ * 0 1
+ *
+ * means that each action taken will reveal information only for itself
+ *
+ * and the all 1's matrix
+ *
+ * 1 1
```

Comment:
is it worthwhile mentioning the input format mapping to this here? or maybe later in some wiki

Reply:
https://github.com/VowpalWabbit/vowpal_wabbit/pull/4392